src/main/network/packet.rs

```diff
@@ -421,364 +831,763 @@ impl PacketDisplay for *const c::Packet {
+        }
+
+        // The TCP header supports 32-bit timestamps; since these are used for network latency
+        // calculations, which are usually on the order of milliseconds, we trade off some precision
+        // here for supporting longer simulations.
+        tcp.header.timestamp = from_legacy_timestamp(ts_val);
+        tcp.header.timestamp_echo = from_legacy_timestamp(ts_echo);
     }
 
     #[no_mangle]
-    pub extern "C-unwind" fn packet_getPayloadSize(packet: *const c::Packet) -> u64 {
-        unsafe { c::legacypacket_getPayloadSize(packet) }
+    pub extern "C-unwind" fn packet_getTCPHeader(packet_ptr: *const Packet) -> c::PacketTCPHeader {
+        let packet = PacketRc::borrow_raw(packet_ptr);
+
+        let IpAddr::V4(src_ip) = packet.header.src else {
+            unimplemented!()
+        };
+        let IpAddr::V4(dst_ip) = packet.header.dst else {
+            unimplemented!()
+        };
+        let Data::LegacyTcp(tcp_rc) = &packet.data else {
+            unimplemented!()
+        };
+        let tcp = tcp_rc.borrow();
+
+        let mut c_hdr: c::PacketTCPHeader = unsafe { MaybeUninit::zeroed().assume_init() };
+
+        c_hdr.flags = tcp_flags_to_legacy_flags(tcp.header.flags);
+        c_hdr.sourceIP = u32::from(src_ip).to_be();
+        c_hdr.sourcePort = tcp.header.src_port.to_be();
+        c_hdr.destinationIP = u32::from(dst_ip).to_be();
+        c_hdr.destinationPort = tcp.header.dst_port.to_be();
+        c_hdr.sequence = tcp.header.sequence;
+        c_hdr.acknowledgment = tcp.header.acknowledgement;
+        c_hdr.selectiveACKs = to_legacy_sel_acks(tcp.header.selective_acks);
+        c_hdr.window = u32::from(tcp.header.window_size);
+        if let Some(scale) = tcp.header.window_scale {
+            c_hdr.windowScale = scale;
+            c_hdr.windowScaleSet = true;
+        }
+        c_hdr.timestampValue = to_legacy_timestamp(tcp.header.timestamp);
+        c_hdr.timestampEcho = to_legacy_timestamp(tcp.header.timestamp_echo);
+
+        c_hdr
     }
 
     #[no_mangle]
-    pub extern "C-unwind" fn packet_getDestinationIP(packet: *const c::Packet) -> libc::in_addr_t {
-        unsafe { c::legacypacket_getDestinationIP(packet) }
+    pub extern "C-unwind" fn packet_setPayloadWithMemoryManager(
+        packet_ptr: *mut Packet,
+        src: UntypedForeignPtr,
+        src_len: u64,
+        mem: *const MemoryManager,
+    ) {
+        // Read data from the managed process into the packet's payload buffer.
+        let packet = PacketRc::borrow_raw_mut(packet_ptr);
+        let mem = unsafe { mem.as_ref() }.unwrap();
+
+        let Data::LegacyTcp(tcp) = &packet.data else {
+            unimplemented!()
+        };
+
+        let len = usize::try_from(src_len).unwrap_or(usize::MAX);
+        let src = ForeignArrayPtr::new(src.cast::<u8>(), len);
+
+        // We want the dst buf on the heap so we don't have to copy it later.
+        let mut dst = vec![0u8; len].into_boxed_slice();
```

Comment:
I think it might be slightly more efficient to use `MaybeUninit` here to avoid zero-filling the vec. Something like `Box::<[u8]>::new_uninit_slice(len)`, writing bytes to the slice from the memory manager, then `dst.assume_init()` to go back to a `Box::<[u8]>`. But maybe just a comment would be good for now.

Reply:
Hmm, do you think zero-filling would cause the performance hit we're seeing in the benchmark?

Reply:
Nevermind, [it didn't make a noticeable difference](https://github.com/shadow/benchmark-results/blob/master/tor/2025-02-12-T21-10-16/plots/run_time.png)